_posts/2024-01-10-semantics-of-regular-expressions.markdown

```diff
@@ -0,0 +1,386 @@
+{
+  OperationalIsCoalgebraHomomorphism<A>();
+  DenotationalIsCoalgebraHomomorphism<A>();
+  UniqueCoalgebraHomomorphism<A>(Operational, Denotational, e);
+}
+```
+
+### Operational Semantics as Algebra Homomorphism
+
+As a little extra, for the sake of symmetry, let us also prove that `Operational` is an algebra homomorphism. (We already know that it is a coalgebra homomorphism, and that `Denotational` is both an algebra and coalgebra homomorphism.)
+
+```
+lemma OperationalIsAlgebraHomomorphism<A(!new)>()
+  ensures IsAlgebraHomomorphism<A>(Operational)
+{...}
+```
+
+The idea of the proof is to take advantage of `Denotational` being an algebra homomorphism, by translating its properties to `Operational` via the lemma in [Denotational and Operational Semantics Are Bisimilar](#denotational-and-operational-semantics-are-bisimilar).  The relevant new statements capture that bisimilarity is symmetric and a congruence with respect to the `Star` operation:
+
+```
+greatest lemma BisimilarityIsSymmetric<A(!new)>[nat](L1: Lang, L2: Lang)
+  ensures Bisimilar(L1, L2) ==> Bisimilar(L2, L1)
+  ensures Bisimilar(L1, L2) <== Bisimilar(L2, L1)
+{}
+
+lemma StarCongruence<A(!new)>(L1: Lang, L2: Lang)
+  requires Bisimilar(L1, L2)
+  ensures Bisimilar(Star(L1), Star(L2))
+{...}
+```
+
+## Conclusion
+
+We have used Dafny’s built-in inductive and coinductive reasoning capabilities to define two language semantics for regular expressions: denotational and operational semantics. Through a number of dualities — construction and deconstruction, algebras and coalgebras, and congruence and bisimilarity — we have proven the semantics to be two sides of the same coin. The blogpost is inspired by research in the field of [Coalgebra](https://en.wikipedia.org/wiki/F-coalgebra), which was pioneered by [Rutten](https://pdf.sciencedirectassets.com/271538/1-s2.0-S0304397500X01466/1-s2.0-S0304397500000566/main.pdf), [Gumm](https://www.researchgate.net/publication/2614339_Elements_Of_The_General_Theory_Of_Coalgebras), and others. The concept of well-behaved semantics goes back to [Turi and Plotkin](https://homepages.inf.ed.ac.uk/gdp/publications/Math_Op_Sem.pdf) and was adapted by [Jacobs](https://link.springer.com/chapter/10.1007/11780274_20) to the case of regular expressions. We heavily used automata theoretic constructions from the 1960s, originally investigated by [Brzozowski](https://dl.acm.org/doi/10.1145/321239.321249) (a more modern presentation can be found e.g. [here](https://alexandrasilva.org/files/thesis.pdf)). Due to space constraints, we had to exclude the details of most of the proofs. To dive deep, please take a look at the full Dafny source code, which is available [here](../../../../assets/src/semantics-of-regular-expressions/Languages.dfy), [here](../../../../assets/src/semantics-of-regular-expressions/Semantics.dfy), and [here](../../../../assets/src/semantics-of-regular-expressions/Expressions.dfy).
```

Comment:
I don't know that it really makes sense to say "due to space constraints" in a blog post. Maybe say something about focusing on the most important intuitive aspects of the proof, instead?